### PR TITLE
Refactor earthquakes to its own class

### PR DESCRIPTION
**Pull request by tobylane**
**Describe what the proposed change does**
- Earthquakes are now ~~a sub class of~~ an instance that belongs to World, with ongoing/next quakes transferred. No additional features. The old recount for current_map_earthquake was faulty so it wasn't copied over (World:afterLoad).

### CODE REVIEW

**Comment by lewri** [comment]
I had originally planned myself to make an `Event` class with `Earthquake`, `VIP`, `Emergency` and `Epidemic` being subclasses but this very likely works too. Maybe better, because I'm sure all those events above might be too diverse to tie to its own overarching class.

File: CorsixTH/Lua/earthquake.lua [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+      if object.strength then
+        object:machineUsed(room)
+      end
+    end
+  end
+
+  self.remaining_damage = self.remaining_damage - 1
+  self.damage_timer = self.damage_timer + damage_time
+
+  -- Patient falling: work in progress, see PR 1848
```
Comment (lewri) [inline CorsixTH/Lua/earthquake.lua:145]: \*tips hat\*

File: CorsixTH/Lua/earthquake.lua [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+--  active (boolean) If we are currently running the warning or damage timers
+--   (after start_day of start_month is passed)
+--  next_planned (boolean) If the next earthquake has been planned/generated
+--  start_month (integer) The month the earthquake warning is triggered
+--  start_day (integer) The day of the month the earthquake warning is triggered
+--  size (integer) The amount of damage the earthquake causes (1-9)
+--  remaining_damage (integer) The amount of damage this earthquake has yet to inflict
+--  damage_timer (integer) The number of hours until the earthquake next inflicts damage if active
+--  warning_timer (integer) The number of hours left until the real damaging earthquake begins
+--  current_map_earthquake (integer) The number of earthquakes that have happened on this map so far
```
Comment (lewri) [inline CorsixTH/Lua/earthquake.lua:37]: 2 space indent the comment if you want that emphasis. Line 29 is out of place though (even if you intended it as a run on of 28)

File: CorsixTH/Lua/cheats.lua [hunk 3]
```diff
@@ -168,6 +168,10 @@ function Cheats:cheatVip()
   self.hospital:createVip()
 end
 
+function Cheats:cheatEarthquake()
+  return self.hospital.world.earthquake:createEarthquake()
+end
+
```
Comment (lewri) [inline CorsixTH/Lua/cheats.lua:174]: Duplicating line 149-153

File: CorsixTH/Lua/hospitals/player_hospital.lua [hunk 4]
```diff
@@ -774,6 +774,47 @@ function PlayerHospital:playSound(sound)
   end
 end
 
+--! The UI parts of earthquake ticks
+--!param stage (string) Stage of the active earthquake
+function PlayerHospital:tickEarthquake(stage)
```
Comment (lewri) [inline CorsixTH/Lua/hospitals/player_hospital.lua:779]: The trouble here is an earthquake is a world event. The announcements aren't though.

File: CorsixTH/Lua/hospitals/player_hospital.lua [hunk 5]
```diff
@@ -774,6 +774,47 @@ function PlayerHospital:playSound(sound)
...
+  local ui = self.world.ui
+  local announcements = {
+    "quake001.wav", "quake002.wav", "quake003.wav", "quake004.wav",
+  }
+
+  -- Start of earhquakes
```
Comment (lewri) [inline CorsixTH/Lua/hospitals/player_hospital.lua:785]:
```suggestion
  -- Start of earthquake
```

**Review by lewri — COMMENTED** [review]
We're definitely moving a lot around, so it's going to be a little challenging to make sure it's all correct.

File: CorsixTH/Lua/hospitals/player_hospital.lua [hunk 6]
```diff
@@ -774,6 +774,47 @@ function PlayerHospital:playSound(sound)
...
+
+  elseif stage == "sound" then
+    -- Play the earthquake sound. It has different names depending on what the language contains.
+    if self.world.app.audio:soundExists("quake2.wav") then
+      ui:playSound("quake2.wav")
+    else
+      ui:playSound("quake.wav")
+    end
+  else
+    print("Unknown stage: " .. stage)
```
[the remark this answers is not in the document]
Comment (tobylane) [reply CorsixTH/Lua/hospitals/player_hospital.lua:814]: This should be an assert, to always crash on bad input, with a fix for nil stage.

File: CorsixTH/Lua/hospitals/player_hospital.lua (hunk 4 above)
Comment (tobylane) [reply CorsixTH/Lua/hospitals/player_hospital.lua:779]: These are the UI parts of the world event, though there may be more from competitor hospitals later.

File: .luacheckrc [hunk 7]
```diff
@@ -34,9 +34,9 @@ globals = { -- Globals
             -- Game classes
             "AIHospital", "AnimationManager", "AnimationEffect", "App", "Audio",
             "CallsDispatcher", "Cheats", "ChildClass", "Command", "Door", "DrawFlags",
-            "DummyRootNode", "EndConditions", "Entity", "EntityMap", "Epidemic",
-            "FileSystem", "FileTreeNode", "FilteredFileTreeNode", "GameUI", "Graphics",
-            "GrimReaper", "Hospital", "Humanoid", "HumanoidRawWalk",
+            "DummyRootNode", "Earthquake","EndConditions", "Entity", "EntityMap",
```
Comment (TheCycoONE) [inline .luacheckrc:37]: nit, missing space after comma.

File: CorsixTH/Lua/earthquake.lua [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  self.damage_timer = self.damage_timer + damage_time
+
+  -- Patient falling: work in progress, see PR 1848
+end
+
+--! Check if it's time for an earthquake
+function Earthquake:onEndDay()
+  if self.world.game_date:monthOfGame() == self.start_month and
+      self.world.game_date:dayOfMonth() == self.start_day then
+    -- Warn the user that an earthquake is on the way
```
Comment (TheCycoONE) [inline CorsixTH/Lua/earthquake.lua:152]: Comment doesn't match code

File: CorsixTH/Lua/world.lua [hunk 9]
```diff
@@ -886,7 +760,7 @@ function World:setSpeed(speed)
   end
   if speed == "Pause" or self.system_pause then
     -- stop screen shaking if there was an earthquake in progress
-    if self.next_earthquake.active then
+    if self.earthquake:isActive() then
       self.ui:endShakeScreen()
```
Comment (TheCycoONE) [inline CorsixTH/Lua/world.lua:764]: You moved the beginShakeScreen to PlayerHospital but not endShakeScreen. I don't have a strong preference for which they're in, but they should be in the same place.

**Comment by tobylane** [comment]
Made changes from feedback.

**Review by TheCycoONE — APPROVED** [review]

File: CorsixTH/Lua/earthquake.lua [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE. --]]
+
+--! Manages all earthquakes in the world
+class "Earthquake"
+
+---@type Earthquake
+local Earthquake = _G["Earthquake"]
+
+-- All fields relating to the current or next earthquake:
```
Comment (lewri) [inline CorsixTH/Lua/earthquake.lua:27]: @tobylane a bit but out of curiosity would a block quote be more sufficient here?

File: CorsixTH/Lua/earthquake.lua [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+            rd = math.max(rd, object.quake_points)
+            object.quake_points = nil
+          end
+        end
+      end
+      old_quake.remaining_damage = rd
+      old_quake.damage_timer = damage_time
+      old_quake.warning_timer = 0
+    end
+  end
```
Comment (lewri) [inline CorsixTH/Lua/earthquake.lua:228]: This is from 0.61... I don't think we would be at any risk of removing this afterload too.

Comment (tobylane) [reply CorsixTH/Lua/earthquake.lua:228]: Now it's done, I'll leave it for a mass clean up.

File: CorsixTH/Lua/earthquake.lua (hunk 10 above)
Comment (tobylane) [reply CorsixTH/Lua/earthquake.lua]: Done and rebased. Ready.

File: CorsixTH/Lua/earthquake.lua (hunk 1 above)
Comment (lewri) [reply CorsixTH/Lua/earthquake.lua:148]: 1848 is now ready, not sure what one should merge first

**Pull request merged**